module/LearnZF2AjaxImageGallery/src/LearnZF2AjaxImageGallery/Controller/IndexController.php

```diff
@@ -0,0 +1,145 @@
+
+use LearnZF2AjaxImageGallery\Form\AjaxImageUploadForm;
+use Zend\Mvc\Controller\AbstractActionController;
+use Zend\View\Model\JsonModel;
+use Zend\View\Model\ViewModel;
+use Zend\Json\Json;
+use Admin\Form\ContentForm;
+use Zend\File\Transfer\Adapter\Http;
+use Zend\Validator\File\IsImage;
+use Zend\Validator\File\Size;
+use Zend\Validator\File\Extension;
+
+class IndexController extends AbstractActionController
+{
+    protected $acceptCriteria = [
+        'Zend\View\Model\JsonModel' => ['application/json'],
+        'Zend\View\Model\ViewModel' => ['text/html'],
+    ];
+
+    public $view;
+
+    public function __construct()
+    {
+        $this->view = new ViewModel();
+    }
+
+    public function indexAction()
+    {
+        $this->acceptableviewmodelselector($this->acceptCriteria);
+        $form = new AjaxImageUploadForm();
+        $this->view->form = $form;
+
+        if ($this->getRequest()->isPost()) {
+            // $form->setInputFilter($content->getInputFilter());
+            $data = array_merge_recursive(
+                $this->getRequest()->getPost()->toArray(),
+                $this->getRequest()->getFiles()->toArray()
+            );
+
+            $form->setData($data);
+            if (!empty($data["imageUpload"]) && $this->getRequest()->isXmlHttpRequest()) {
+                return $this->uploadImages();
+            }
+        }
+        return $this->view;
+    }
+
+    /**
+     * Get all files from all folders and list them in the gallery
+     */
+    public function filesAction()
+    {
+        $this->view->setTerminal(true);
+        $dir = new \RecursiveDirectoryIterator('public/userfiles/', \FilesystemIterator::SKIP_DOTS);
+        $it  = new \RecursiveIteratorIterator($dir, \RecursiveIteratorIterator::SELF_FIRST);
+        $it->setMaxDepth(3);
+        $files = [];
+        $i = 0;
+        foreach ($it as $file) {
+            if ($file->isFile()) {
+                $filepath = explode("public", $file->getRealPath()); // ugly workaround :(
+                $files[$i]["link"] = Json::encode($filepath[1]);
+                $files[$i]["filename"] = Json::encode($file->getFilename());
+                $i++;
+            }
+        }
+        return new JsonModel($files);
+    }
+
+    /**
+     * Create directories
+     * @return  void
+     */
+    private function createDirectories()
+    {
+        if (!is_dir('public/userfiles/')) {
+            mkdir('public/userfiles/');
+        }
+        if (!is_dir('public/userfiles/'.date("Y_M"))) {
+            mkdir('public/userfiles/'.date("Y_M"));
+        }
+        if (!is_dir('public/userfiles/'.date("Y_M").'/images/')) {
+            mkdir('public/userfiles/'.date("Y_M").'/images/');
+        }
+    }
+
+    /**
+     * Upload all images async
+     *
+     * @param  array $files
+     * @return Response containing headers with information about each image
+     */
+    private function uploadImages()
+    {
+        $this->view->setTerminal(true);

```

Comment:
iirc, if we are using <code>acceptableviewmodelselector</code> plugin, we don't need setTerminal() as far as the <code>ContentType</code> and <code>Accept</code> header is <code>application/json</code>